#DADA2
Following the DADA2 pipeline from benjjneb https://benjjneb.github.io/dada2/tutorial.html

```{r}
#installing the packages 
install.packages("tidyverse")
install.packages("Rtools")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("dada2", version = "3.20", ask = T)
```

```{r}
if(!requireNamespace("BiocManager")){
  install.packages("BiocManager")
}
BiocManager::install("phyloseq")

```

```{r}
#loading the packages
library(tidyverse)
library(dada2)
```

```{r}
Sys.setlocale("LC_CTYPE", "en_US.UTF-8")
```


```{r}
#uploading my data
path <- "DNA data" 
list.files(path)
```

```{r}
fnFs <- sort(list.files(path, pattern="_R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq.gz", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), function(x) paste(x[1:3], collapse="_"))
```

```{r}
fnFs
```

```{r}
fnRs
```



```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```



#Removing Primers
```{r}
#Removing 25 nucleotides (primer length was 21)
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(275,275),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

#calculating errors 

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)

``` 

```{r}
errR <- learnErrors(filtRs, multithread=TRUE)

```


```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
```
```{r}
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
```
```{r}
dadaFs[[1]]

```


```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```


```{r}
table(nchar(getSequences(seqtab)))

```

##Removing chimeras
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
```
##Removing ASVs with less then 10 total reads across all samples
```{r}
asv_counts<-colSums(seqtab.nochim)
seqtab.filtered<-seqtab.nochim[,asv_counts > 10]
```


```{r}
sum(seqtab.nochim)/sum(seqtab)
sum(seqtab.filtered)/sum(seqtab.nochim)

```

##some stats
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

##Road to files
```{r}
getwd()

```

```{r}
save.image("enviroment_pc_lab.rdata")



```
#Assigninng taxonomy 
```{r}
taxa <- assignTaxonomy(seqtab.filtered, 
                        "C:/Users/pontusfe/OneDrive - Universitetet i Agder/PhD Norge/Students in Research/Aurora/pr2_version_5.1.0_SSU_dada2.fasta.gz", 
                        multithread=TRUE, 
                        taxLevels = c("Domain", "Supergroup", "Division", "Subdivision", "Class", "Order", "Family", "Genus", "Species"),
                        outputBootstraps = TRUE)
```


```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

```{r}
write.csv(taxa, "taxa_assignment2.csv", row.names=TRUE)
saveRDS(taxa, "taxa_assignment2.rds")
```

```{r}
save.image("enviroment_pc_lab.rdata")
```

#Handoff to Phyloseq

```{r}
samples.out <- rownames(seqtab.filtered)
split_names <- strsplit(samples.out, "_")
Location <- sapply(split_names, `[`, 1)  
Date <- sapply(split_names, `[`, 2)       
Depth <- sapply(split_names, `[`, 3)
Day <- substr(Date, 1, 2)
Month <- substr(Date, 3, 4)
Year <- paste0("20", substr(Date, 5, 6))  
Season <- ifelse(Month %in% c("12", "01", "02"), "Early_winter",
          ifelse(Month %in% c("03", "04", "05"), "Post_Spring_bloom",
          ifelse(Month %in% c("06", "07", "08"), "Summer", "Early_Fall")))
samdf <- data.frame(Location=Location, Date=paste(Year, Month, Day, sep="-"),Depth=Depth, Season=Season)
rownames(samdf) <- samples.out
head(samdf)

```
###Filtrerer ut bootstraps<80
```{r}
taxa_tax <- as.data.frame(taxa$tax)
taxa_boot <- as.data.frame(taxa$boot)

```

```{r}
bootstrap_min <- 80
high_confidence <- which(taxa_boot$Species >= bootstrap_min)

# Filtrert taksonomi
taxa_filtered <- taxa_tax[high_confidence, ]

# Og sekvenstabellen må filtreres tilsvarende – altså beholde de samme ASV-ene
seqtab_filtered_boot <- seqtab.filtered3[, high_confidence]

```


```{r}
ps <- phyloseq(
  otu_table(seqtab_filtered_boot, taxa_are_rows=FALSE),
  sample_data(samdf),
  tax_table(as.matrix(taxa_filtered))
)

```
##Save
```{r}
saveRDS(ps, "phyloseqtable.rds")

```

```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```
##Removing metazoa and Fungi
```{r}
ps_protists <- subset_taxa(ps, !(Subdivision %in% c("Metazoa", "Fungi")))
table(tax_table(ps_protists)[, "Subdivision"])

```

making a dataframe
```{r}
ps_protist_df <- psmelt(ps_protists)
```
#Rarefying 
```{r}
ps_protists_rarefied <- rarefy_even_depth(ps_protists, rngseed = 123, verbose = TRUE)

```

```{r}
#making a dataframe
df_protist_rarefied <- psmelt(ps_protists_rarefied)
#adding a column with relative abundance
df_protist_rarefied <- df_protist_rarefied %>%
  group_by(Sample) %>%
  mutate(Relative_Abundance = Abundance / sum(Abundance))  
#making an OTU table 
asv_table_rarefied <- as(otu_table(ps_protists_rarefied), "matrix")
#Removing reads below 5 
asv_table_rarefied[asv_table_rarefied < 5] <- 0
asv_table_rarefied <- asv_table_rarefied[, colSums(asv_table_rarefied) > 0]
```

```{r}
asv_table <- as(otu_table(ps_protists), "matrix")

# Rarefaction curves
rarecurve(asv_table, step = 100, col = rainbow(nrow(asv_table)), label = TRUE, cex=0.2)
```

```{r}
#making OTU table
otu_table_rarefied <- otu_table(ps_protists_rarefied) 
ps_protist_rarefied_df <- psmelt(ps_protists_rarefied)
#number of sequences left
total_seqs <- sum(otu_table_rarefied)
print(total_seqs)
```

```{r}
save.image(file = "DNA_Environment.RData")
```

